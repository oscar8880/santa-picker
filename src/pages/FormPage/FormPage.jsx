import styles from './FormPage.module.scss';
import React, { useState } from 'react';
import Text from '../../components/Text/Text';
import PageWrapper from '../../components/PageWrapper/PageWrapper';
import FormikTextInput from '../../components/FormikTextInput/FormikTextInput';
import { Container, Row, Col } from 'react-grid-system';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
import FormikCheckbox from '../../components/FormikCheckbox/FormikCheckbox';

const FormPage = () => {
  const [numParticipants, setNumParticipants] = useState(3);
  const handleSubmit = (values) => {
    console.log(values);
  };

  const generateValidationSchema = () => {
    const schema = {};
    for (let i = 0; i < numParticipants; i++) {
      schema[`participantName${i}`] = Yup.string()
        .required('Required')
        .min(2, 'Must be at least two characters')
        .max(30, 'Must be 30 characters or fewer');
      schema[`participantEmail${i}`] = Yup.string()
        .email('Invalid email address')
        .required('Required')
        .max(100, 'Must be 100 characters or fewer');
    }
    return Yup.object().shape(schema);
  };

  const generateInitialValues = () => {
    const values = {};
    for (let i = 0; i < numParticipants; i++) {
      values[`participantName${i}`] = '';
      values[`participantEmail${i}`] = '';
    }
    return values;
  };

  const generateFormFields = () => {
    const fields = [];
    for (let i = 0; i < numParticipants; i++) {
      fields.push(
        <Row key={`participantKey${i}`}>
          <div className={styles.FormPage__InputRow}>
            <Col xs={12} sm={6}>
              <div className={styles.FormPage__Input}>
                <FormikTextInput
                  label="Name"
                  name={`participantName${i}`}
                  id={`participantName${i}`}
                  type="text"
                />
              </div>
            </Col>
            <Col xs={12} sm={6}>
              <div className={styles.FormPage__Input}>
                <FormikTextInput
                  label="Email"
                  name={`participantEmail${i}`}
                  id={`participantEmail${i}`}
                  type="text"
                />
              </div>
            </Col>
          </div>
        </Row>,
      );
    }
    fields.push(
      <FormikCheckbox
        key="acceptedTerms"
        id="acceptedTerms"
        name="acceptedTerms"
      >
        Add a spending limit?
      </FormikCheckbox>,
    );
    return fields;
  };

  return (
    <div className={styles.FormPage}>
      <PageWrapper>
        <Container fluid>
          <Row>
            <Col xs={12}>
              <div className={styles.FormPage__Title}>
                <Text tagName="h1" appearance="h1">
                  Participant details
                </Text>
              </div>
              <div className={styles.FormPage__BlurbText}>
                <Text tagName="p" appearance="h5">
                  Fill out the name and email address of each person taking part
                </Text>
              </div>
            </Col>
          </Row>
          <Formik
            initialValues={generateInitialValues()}
            validationSchema={generateValidationSchema()}
            onSubmit={handleSubmit}
          >
            <Form>
              {generateFormFields()}
              <button
                type="button"
                onClick={() => setNumParticipants(numParticipants + 1)}
              >
                Add
              </button>
              <button type="submit">Submit</button>
            </Form>
          </Formik>
          <div className={styles.FormPage__BlurbText}>
            <Text tagName="p" appearance="h5">
              None of this information is saved or used other than to send out
              the Santa allocations
            </Text>
          </div>
        </Container>
      </PageWrapper>
    </div>
  );
};

export default FormPage;
