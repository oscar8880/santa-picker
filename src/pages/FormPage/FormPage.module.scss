@import "../../scss/functions";
@import "../../scss/variables";

.FormPage {
  display: flex;
  flex-direction: column;
  padding-bottom: $spacer-12;
  padding-top: $spacer-10;
  @media (min-width: $breakpoint-md) {
    padding-top: $spacer-12;
  }
}

.FormPage__Title {
  padding-bottom: $spacer-4;
}

.FormPage__BlurbText {
  color: $grey-500;
  padding-bottom: $spacer-8;
}

.FormPage__InputRow {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
}

.FormPage__Input {
  padding-bottom: $spacer-8;
}

.FormPage__CTA {
  display: flex;
  justify-content: center;
  padding-bottom: $spacer-8;

  @media (min-width: $breakpoint-md) {
    justify-content: flex-start;
  }
}
