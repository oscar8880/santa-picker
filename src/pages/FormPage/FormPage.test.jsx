import React from 'react';
import { render } from '@testing-library/react';

import FormPage from './FormPage';

test('matches snapshot', async () => {
  const { asFragment } = render(<FormPage />);

  expect(asFragment()).toMatchSnapshot();
});
