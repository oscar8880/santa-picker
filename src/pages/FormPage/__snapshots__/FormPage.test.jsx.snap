// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormPage"
  >
    <div
      class="PageWrapper"
    >
      <div
        style="box-sizing: border-box; position: relative; margin-left: auto; margin-right: auto; padding-left: 15px; padding-right: 15px;"
      >
        <div
          style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
        >
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 100%; flex-basis: 100%; flex-grow: 0; flex-shrink: 0; max-width: 100%; margin-left: 0%;"
          >
            <div
              class="FormPage__Title"
            >
              <h1
                class="Text Text__h1"
              >
                Participant details
              </h1>
            </div>
            <div
              class="FormPage__BlurbText"
            >
              <p
                class="Text Text__h5"
              >
                Fill out the name and email address of each person taking part
              </p>
            </div>
          </div>
        </div>
        <form
          action="#"
        >
          <div
            style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
          >
            <div
              class="FormPage__InputRow"
            >
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantName0"
                    >
                      Name
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantName0"
                      name="participantName0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantEmail0"
                    >
                      Email
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantEmail0"
                      name="participantEmail0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
          >
            <div
              class="FormPage__InputRow"
            >
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantName1"
                    >
                      Name
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantName1"
                      name="participantName1"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantEmail1"
                    >
                      Email
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantEmail1"
                      name="participantEmail1"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start;"
          >
            <div
              class="FormPage__InputRow"
            >
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantName2"
                    >
                      Name
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantName2"
                      name="participantName2"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
              >
                <div
                  class="FormPage__Input"
                >
                  <div
                    class="FormikTextInput__Container"
                  >
                    <label
                      class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
                      for="participantEmail2"
                    >
                      Email
                    </label>
                    <input
                      class="FormikTextInput__Input"
                      id="participantEmail2"
                      name="participantEmail2"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <label
            class="checkbox"
          >
            <input
              id="acceptedTerms"
              name="acceptedTerms"
              type="checkbox"
              value=""
            />
            Add a spending limit?
          </label>
          <button
            type="button"
          >
            Add
          </button>
          <button
            type="submit"
          >
            Submit
          </button>
        </form>
        <div
          class="FormPage__BlurbText"
        >
          <p
            class="Text Text__h5"
          >
            None of this information is saved or used other than to send out the Santa allocations
          </p>
        </div>
        <span
          style="display: table; clear: both;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
